### PR TITLE
libocfs2: roll-back when dir_index creation fails

### PR DESCRIPTION
**Pull request by ganghe**
When we try to create a directory index tree, we should truncate
the directory index tree if this creation failed for some reason,
e.g. there is not any more blocks to allocate.
Otherwise, the file was attached a problematic directory index
tree, this index will affect file lookup under this directory file.

### CODE REVIEW

**Comment by ganghe** [comment]
I'd like to invite more people to review this patch, before it is merged.

Thanks
Gang

**Comment by josephhz** [comment]
Hi Gang, in which case do you encounter this issue?

**Comment by ganghe** [comment]
> Hi Gang, in which case do you encounter this issue?

Hi Joseph,

This problem was exposed during running ocfs2-test test cases, the reproduce command line is as below,
"discontig_runner.sh -m node1,node2,node3 -f 0 -d /dev/vdb1 -s pcmk -n hacluster /mnt/ocfs2"
The detailed situation is, 
The file system was full, then fsck.ocfs2 was run on this file system, the fsck.ocfs2 tried to create a dir index tree for a directory file, it failed since there were not more blocks to allocate.
But, the problem was, the index tree of this directory file was broken, it affected file lookup.
If this directory file is a system directory file, the case is very bad, the file system can not be mounted.

e.g.
2021/10/27,04:36:06  Do fsck.ocfs2
fsck.ocfs2 1.8.7
Checking OCFS2 filesystem in /dev/vdb1:
  Label:              ocfs2-discontig-bg-tests-211027025343-multi-nodes
  UUID:               E8C941598C454A6F86059E57B1328743
  Number of blocks:   167772160
  Block size:         512
  Number of clusters: 20971520
  Cluster size:       4096
  Number of slots:    3

/dev/vdb1 was run with -f, check forced.
Pass 0a: Checking cluster allocation chains
Pass 0b: Checking inode allocation chains
Pass 0c: Checking extent block allocation chains
Pass 1: Checking inodes and blocks
[INODE_SPARSE_SIZE] Inode 130867 has a size of 0 but has 157892608 blocks of actual data. Correct the file size? y
Pass 2: Checking directory entries
[DX_TREE_MISSING] Directory 18 is missing index. Rebuild? y
pass2: Unable to find available bit while rebuild indexed dirs.
fsck.ocfs2: Unable to find available bit while performing pass 2
2021/10/27,04:36:32  fsck failed
2021/10/27,04:36:32  [*] Mount /dev/vdb1 to /mnt/ocfs2:
2021/10/27,04:36:32  /usr/bin/sudo -u root /bin/mount -t ocfs2 -o localalloc=0 /dev/vdb1 /mnt/ocfs2
mount.ocfs2: Internal logic failure while mounting /dev/vdb1 on /mnt/ocfs2. Check 'dmesg' for more information on this error 22.
2021/10/27,04:36:32  Mount volume failed

File: libocfs2/dir_indexed.c [hunk 1]
```diff
@@ -1262,10 +1262,10 @@ errcode_t ocfs2_dx_dir_build(ocfs2_filesys *fs,
 
 	ret = ocfs2_write_dx_root(fs, dr_blkno, dx_buf);
 	if (ret)
-		goto out;
+		goto trunc_out;
```
Comment (josephhz) [inline libocfs2/dir_indexed.c:1265]: Here we still haven't written dx_buf, why need truncate back?

**Comment by ganghe** [comment]
I will send this patch for v2.

**Pull request closed without merging**